lib/index.js

```diff
@@ -411,25 +432,59 @@ elasticlunr.Index.prototype.fieldSearch = function (queryTokens, fieldName, conf
 
         var score = tf * idf * fieldLengthNorm * penality;
 
-        if (docRef in scores) {
-          scores[docRef] += score;
+        if (docRef in queryTokenScores) {
+          queryTokenScores[docRef] += score;
         } else {
-          scores[docRef] = score;
+          queryTokenScores[docRef] = score;
         }
       }
     }, this);
+    
+    scores = this.mergeScores(scores, queryTokenScores, booleanType);
   }, this);
 
-  if (booleanType == 'AND') {
-    scores = this.intersect(scores, docTokens, queryTokens.length);
-  }
-
   scores = this.coordNorm(scores, docTokens, queryTokens.length);
-
   return scores;
 };
 
 /**
+ * Merge the scores from one set of tokens into an accumulated score table.
+ * Exact operation depends on the op parameter. If op is 'AND', then only the
+ * intersection of the two score lists is retained. Otherwise, the union of
+ * the two score lists is returned. For internal use only.
+ *
+ * @param {Object} bool accumulated scores. Should be null on first call.
+ * @param {String} scores new scores to merge into accumScores.
+ * @param {Object} op merge operation (should be 'AND' or 'OR').
+ *
+ */
+
+elasticlunr.Index.prototype.mergeScores = function (accumScores, scores, op) {
+    if (!accumScores) {
+        return scores; 
+    }
+    if (op == 'AND') {
+        var intersection = {};
+        for (var docRef in scores) {
+            if (docRef in accumScores) {
+                intersection[docRef] = accumScores[docRef] + scores[docRef];
+            }
+        }
```

Comment:
Would this give some uncorrect results? for example, query tokens contains ["A", "B", "C"], but files only contain "A" and "B" would also be returned?
Sorry I have not enough time to understand your code. I will look it later. Add some comments would be great.